tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -848,3 +848,75 @@ def prepare_r_uls_r_lus(
             r_lu_matrix,
         )
         # TODO: beta sobolev needs to be recalculated for each iteration, because it depends on number density
+
+    @staticmethod
+    def create_coll_exc_deexc_matrix(
+        coll_exc_coefficient,
+        coll_deexc_coefficient,
+        number_of_levels,
+    ):
+        """Generates a coefficient matrix from collisional excitation/deexcitation coefficients.
+
+        Needs to be multiplied by electron density when added to the overall rate_matrix.
+        Parameters
+        ----------
+        coll_exc_coefficient : pandas.DataFrame
+            DataFrame of collisional excitation coefficients for current (atomic number, ion_number)
+        coll_deexc_coefficient : pandas.DataFrame
+            DataFrame of collisional deexcitation coefficients for (atomic number, ion_number)
+        number_of_levels : int
+            Number of levels for the current atomic number, ion number.
+
+        Returns
+        -------
+        coeff_matrix : np.array (number of levels, number of levels)
+            Square matrix constructed by collisional exc./deexc. coefficients.
+        """
+        diagonal_exc = np.zeros(number_of_levels)
+        diagonal_deexc = np.zeros(number_of_levels)
+        diagonal_exc[
+            coll_exc_coefficient.index.get_level_values(
+                "level_number_lower"
+            ).unique()
+        ] = (
+            -1
+            * coll_exc_coefficient.groupby("level_number_lower")
+            .sum()
+            .values.flatten()
```

Comment:
In principle, `coll_exc_coefficient` should be a Series since it is one column of a DataFrame (corresponding to a specific shell). The `.flatten()` should then not be necessary.